File: AdvantageFramework.Core/Database/Procedures/Job.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Linq;

namespace AdvantageFramework.Core.Database.Procedures
{
    public static partial class Job
    {
        #region  Constants 



        #endregion

        #region  Enum 



        #endregion

        #region  Variables 



        #endregion

        #region  Properties 



        #endregion

        #region  Methods 

        #region   Core Entities 

        public static IEnumerable<AdvantageFramework.Core.Database.Core.Job> LoadCore(System.Linq.IQueryable<Database.Entities.Job> Jobs)
        {
            IEnumerable<AdvantageFramework.Core.Database.Core.Job> LoadCoreRet = default;
            try
            {
                LoadCoreRet = Jobs.Select(Entity => new
                {
                    Entity.JobNumber,
                    Entity.OfficeCode,
                    Entity.ClCode,
                    Entity.DivCode,
                    Entity.PrdCode,
                    Entity.JobDesc,
                    Entity.CompOpen
                }).Select(Entity => new AdvantageFramework.Core.Database.Core.Job()
                {
                    Number = Entity.JobNumber,
                    OfficeCode = Entity.OfficeCode,
                    ClientCode = Entity.ClCode,
                    DivisionCode = Entity.DivCode,
                    ProductCode = Entity.PrdCode,
                    Description = Entity.JobDesc,
                    IsOpen = Entity.CompOpen

                }).ToList();
            }
            catch (Exception)
            {
                LoadCoreRet = null;
            }

            return LoadCoreRet;
        }

        public static IEnumerable<AdvantageFramework.Core.Database.Core.Job> LoadCore(AdvantageFramework.Core.Database.DbContext DbContext)
        {
            IEnumerable<AdvantageFramework.Core.Database.Core.Job> LoadCoreRet = default;
            try
            {
                LoadCoreRet = LoadCore(Load(DbContext));
            }
            catch (Exception ex)
            {
                LoadCoreRet = null;
            }

            return LoadCoreRet;
        }

        #endregion

        public static System.Linq.IQueryable<Database.Entities.Job> LoadAllOpenByClientCode(Database.DbContext DbContext, string ClientCode)
        {
            System.Linq.IQueryable<Database.Entities.Job> LoadAllOpenByClientCodeRet = default;

            LoadAllOpenByClientCodeRet = from Job in LoadAllOpen(DbContext)
                                         where Job.ClCode == ClientCode
                                         select Job;
            return LoadAllOpenByClientCodeRet;
        }

//        public static Global.System.Data.Entity.Infrastructure.DbQuery LoadByUserForEmployeeCode(AdvantageFramework.Core.Database.DbContext DbContext, AdvantageFramework.Security.Database.DbContext SecurityDbContext, string UserCode, string EmployeeCode, bool OpenByJobProcessNumberOnly)
//        {
//            Global.System.Data.Entity.Infrastructure.DbQuery LoadByUserForEmployeeCodeRet = default;
//            int[] ExcludedJobProcessControlNumbers = null;
//            List<AdvantageFramework.Security.Database.Entities.UserClientDivisionProductAccess> UserClientDivisionProductAccessList = null;
//            string[] EmployeeUserCodes = null;
//            int[] JobNumbers = null;
//            ExcludedJobProcessControlNumbers = new int[] { 2, 5, 7, 10, 11, 6, 12 };
//            try
//            {
//                try
//                {
//                    ;
//#error Cannot convert AssignmentStatementSyntax - see comment for details
//                    /* Cannot convert ParenthesizedExpressionSyntax, System.InvalidCastException: Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.Syntax.EmptyStatementSyntax' to type 'Microsoft.CodeAnalysis.CSharp.Syntax.ExpressionSyntax'.
//                       at ICSharpCode.CodeConverter.CSharp.CommentConvertingVisitorWrapper.ConvertHandledAsync[T](SyntaxNode vbNode, SourceTriviaMapKind sourceTriviaMap) in D:\GitWorkspace\CodeConverter\CodeConverter\CSharp\CommentConvertingVisitorWrapper.cs:line 47

//                    Input: (From Entity In AdvantageFramework.Security.Database.Procedures.User.LoadByEmployeeCode(SecurityDbContext, EmployeeCode)
//                                                             Select [UC] = Entity.UserCode)
//                    Context:

//                                        EmployeeUserCodes = (From Entity In AdvantageFramework.Security.Database.Procedures.User.LoadByEmployeeCode(SecurityDbContext, EmployeeCode)
//                                                             Select [UC] = Entity.UserCode).ToArray

//                     */
//                }
//                catch (Exception ex)
//                {
//                    EmployeeUserCodes = null;
//                }

//                if (EmployeeUserCodes is object)
//                {
//                    if (EmployeeUserCodes.Contains(UserCode))
//                    {
//                        EmployeeUserCodes = new[] { UserCode };
//                    }

//                    try
//                    {
//                        UserClientDivisionProductAccessList = (from Entity in AdvantageFramework.Security.Database.Procedures.UserClientDivisionProductAccess.Load(SecurityDbContext)
//                                                               where EmployeeUserCodes.Contains(Entity.UserCode)
//                                                               select Entity).ToList;
//                    }
//                    catch (Exception ex)
//                    {
//                        UserClientDivisionProductAccessList = null;
//                    }
//                }

//                JobNumbers = (from Entity in AdvantageFramework.Core.Database.Procedures.JobComponent.Load(DbContext)
//                              where ExcludedJobProcessControlNumbers.Contains(Entity.JobProcessNumber) == false || ExcludedJobProcessControlNumbers.Contains(Entity.JobProcessNumber) == (OpenByJobProcessNumberOnly == false ? true : false)
//                              select Entity.JobNumber).Distinct.ToArray;
//                if (UserClientDivisionProductAccessList is object && UserClientDivisionProductAccessList.Count > 0)
//                {
//                    JobNumbers = (from Entity in (from Job in LoadAllOpen(DbContext)
//                                                  select new
//                                                  {
//                                                      Job.ClientCode,
//                                                      Job.DivisionCode,
//                                                      Job.ProductCode,
//                                                      Job.Number
//                                                  }).ToList
//                                  where UserClientDivisionProductAccessList.Any(UserAccess => Operators.ConditionalCompareObjectEqual(UserAccess.ClientCode, Entity.ClientCode, false) && Operators.ConditionalCompareObjectEqual(UserAccess.DivisionCode, Entity.DivisionCode, false) && Operators.ConditionalCompareObjectEqual(UserAccess.ProductCode, Entity.ProductCode, false)) && JobNumbers.Contains(Entity.Number)
//                                  select Entity.Number).ToArray;
//                }

//                LoadByUserForEmployeeCodeRet = from Job in LoadAllOpen(DbContext)
//                                               where JobNumbers.Contains(Job.Number)
//                                               select Job;
//            }
//            catch (Exception ex)
//            {
//                LoadByUserForEmployeeCodeRet = default;
//            }

//            return LoadByUserForEmployeeCodeRet;
//        }

        //public static Global.System.Data.Entity.Infrastructure.DbQuery LoadByUserCode(AdvantageFramework.Core.Database.DbContext DbContext, AdvantageFramework.Security.Database.DbContext SecurityDbContext, string UserCode)
        //{
        //    Global.System.Data.Entity.Infrastructure.DbQuery LoadByUserCodeRet = default;
        //    List<AdvantageFramework.Security.Database.Entities.UserClientDivisionProductAccess> UserClientDivisionProductAccess = null;
        //    int[] JobNumbers = null;
        //    string SQL = "";
        //    try
        //    {
        //        UserClientDivisionProductAccess = AdvantageFramework.Security.Database.Procedures.UserClientDivisionProductAccess.LoadByUserCode(SecurityDbContext, UserCode).ToList;
        //        if (UserClientDivisionProductAccess.Count > 0)
        //        {
        //            SQL = string.Format("SELECT JOB_NUMBER FROM dbo.JOB_LOG JL " + "JOIN dbo.SEC_CLIENT SC ON SC.CL_CODE = JL.CL_CODE AND SC.DIV_CODE = JL.DIV_CODE AND SC.PRD_CODE = JL.PRD_CODE " + "WHERE SC.[USER_ID] = '{0}'", UserCode);
        //            try
        //            {
        //                JobNumbers = DbContext.Database.SqlQuery<int>(SQL).ToArray;
        //            }
        //            catch (Exception ex)
        //            {
        //                JobNumbers = null;
        //            }

        //            LoadByUserCodeRet = from Job in DbContext.Jobs.AsQueryable()
        //                                where JobNumbers.Contains(Job.Number)
        //                                select Job;
        //        }
        //        else
        //        {
        //            LoadByUserCodeRet = Load(DbContext);
        //        }
        //    }
        //    catch (Exception ex)
        //    {
        //        LoadByUserCodeRet = default;
        //    }

        //    return LoadByUserCodeRet;
        //}

        //public static Global.System.Data.Entity.Infrastructure.DbQuery LoadOpenByUserCode(AdvantageFramework.Core.Database.DbContext DbContext, AdvantageFramework.Security.Database.DbContext SecurityDbContext, string UserCode)
        //{
        //    Global.System.Data.Entity.Infrastructure.DbQuery LoadOpenByUserCodeRet = default;
        //    List<AdvantageFramework.Security.Database.Entities.UserClientDivisionProductAccess> UserClientDivisionProductAccess = null;
        //    int[] JobNumbers = null;
        //    string SQL = "";
        //    try
        //    {
        //        UserClientDivisionProductAccess = AdvantageFramework.Core.Security.Database.Procedures.UserClientDivisionProductAccess.LoadByUserCode(SecurityDbContext, UserCode).ToList;
        //        if (UserClientDivisionProductAccess.Count > 0)
        //        {
        //            SQL = string.Format("SELECT JOB_NUMBER FROM dbo.JOB_LOG JL " + "JOIN dbo.SEC_CLIENT SC ON SC.CL_CODE = JL.CL_CODE AND SC.DIV_CODE = JL.DIV_CODE AND SC.PRD_CODE = JL.PRD_CODE " + "WHERE SC.[USER_ID] = '{0}'", UserCode);
        //            try
        //            {
        //                JobNumbers = DbContext.Database.SqlQuery<int>(SQL).ToArray;
        //            }
        //            catch (Exception ex)
        //            {
        //                JobNumbers = null;
        //            }

        //            LoadOpenByUserCodeRet = from Job in LoadAllOpen(DbContext)
        //                                    where JobNumbers.Contains(Job.JobNumber)
        //                                    select Job;
        //        }
        //        else
        //        {
        //            LoadOpenByUserCodeRet = LoadAllOpen(DbContext);
        //        }
        //    }
        //    catch (Exception ex)
        //    {
        //        LoadOpenByUserCodeRet = default;
        //    }

        //    return LoadOpenByUserCodeRet;
        //}

        public static System.Linq.IQueryable<Database.Entities.Job> LoadByClientCode(Database.DbContext DbContext, string ClientCode)
        {
            System.Linq.IQueryable<Database.Entities.Job> LoadByClientCodeRet = default;
            LoadByClientCodeRet = from Job in DbContext.Jobs.AsQueryable()
                                  where Job.ClCode == ClientCode
                                  select Job;
            return LoadByClientCodeRet;
        }

        public static System.Linq.IQueryable<Database.Entities.Job> LoadByClientAndDivisionCode(Database.DbContext DbContext, string ClientCode, string DivisionCode)
        {
            System.Linq.IQueryable<Database.Entities.Job> LoadByClientAndDivisionCodeRet = default;
            LoadByClientAndDivisionCodeRet = from Job in DbContext.Jobs.AsQueryable()
                                             where Job.ClCode == ClientCode && Job.DivCode == DivisionCode
                                             select Job;
            return LoadByClientAndDivisionCodeRet;
        }

        public static System.Linq.IQueryable<Database.Entities.Job> LoadByClientAndDivisionAndProductCode(Database.DbContext DbContext, string ClientCode, string DivisionCode, string ProductCode)
        {
            System.Linq.IQueryable<Database.Entities.Job> LoadByClientAndDivisionAndProductCodeRet = default;
            LoadByClientAndDivisionAndProductCodeRet = from Job in DbContext.Jobs.AsQueryable()
                                                       where Job.ClCode == ClientCode && Job.DivCode == DivisionCode && Job.PrdCode == ProductCode
                                                       select Job;
            return LoadByClientAndDivisionAndProductCodeRet;
        }

        public static System.Linq.IQueryable<Database.Entities.Job> LoadByOfficeCode(AdvantageFramework.Core.Database.DbContext DbContext, string OfficeCode)
        {
            System.Linq.IQueryable<Database.Entities.Job> LoadByOfficeCodeRet = default;
            LoadByOfficeCodeRet = from Job in DbContext.Jobs.AsQueryable()
                                  where Job.OfficeCode == OfficeCode
                                  select Job;
            return LoadByOfficeCodeRet;
        }

        public static AdvantageFramework.Core.Database.Entities.Job LoadByJobID(AdvantageFramework.Core.Database.DbContext DbContext, int JobID)
        {
            AdvantageFramework.Core.Database.Entities.Job LoadByJobIDRet = default;
            try
            {
                LoadByJobIDRet = (from Job in DbContext.Jobs.AsQueryable()
                                  where Job.Rowid == JobID
                                  select Job).SingleOrDefault();
            }
            catch (Exception )
            {
                LoadByJobIDRet = default;
            }

            return LoadByJobIDRet;
        }

        public static AdvantageFramework.Core.Database.Entities.Job LoadByJobNumber(AdvantageFramework.Core.Database.DbContext DbContext, int JobNumber)
        {
            AdvantageFramework.Core.Database.Entities.Job LoadByJobNumberRet = default;
            try
            {
                LoadByJobNumberRet = (from Job in DbContext.Jobs.AsQueryable()
                                      where Job.JobNumber == JobNumber
                                      select Job).SingleOrDefault();
            }
            catch (Exception )
            {
                LoadByJobNumberRet = default;
            }

            return LoadByJobNumberRet;
        }

        public static System.Linq.IQueryable<Database.Entities.Job> LoadBySalesClassAndFormatCode(AdvantageFramework.Core.Database.DbContext DbContext, string SalesClassCode, string SalesClassFormatCode)
        {
            System.Linq.IQueryable<Database.Entities.Job> LoadBySalesClassAndFormatCodeRet = default;
            try
            {
                LoadBySalesClassAndFormatCodeRet = (from Job in DbContext.Jobs.AsQueryable()
                                                    where Job.ScCode == SalesClassCode && Job.FormatScCode == SalesClassFormatCode
                                                    select Job);
            }
            catch (Exception )
            {
                LoadBySalesClassAndFormatCodeRet = default;
            }

            return LoadBySalesClassAndFormatCodeRet;
        }

        public static System.Linq.IQueryable<Database.Entities.Job> LoadAllOpen(AdvantageFramework.Core.Database.DbContext DbContext)
        {
            System.Linq.IQueryable<Database.Entities.Job> LoadAllOpenRet = default;
            LoadAllOpenRet = from Job in DbContext.Jobs.AsQueryable()
                             where Job.CompOpen == 1
                             select Job;
            return LoadAllOpenRet;
        }

        public static Microsoft.EntityFrameworkCore.DbSet<Database.Entities.Job> Load(AdvantageFramework.Core.Database.DbContext DbContext)
        {
            return DbContext.Jobs;
        }

        //public static bool Update(AdvantageFramework.Core.Database.DbContext DbContext, AdvantageFramework.Core.Database.Entities.Job Job)
        //{
        //    bool UpdateRet = default;

        //    // objects
        //    bool Updated = false;
        //    bool IsValid = true;
        //    string ErrorText = "";
        //    try
        //    {
        //        DbContext.UpdateObject(Job);
        //        ErrorText = Job.ValidateEntity(IsValid);
        //        if (IsValid)
        //        {
        //            DbContext.SaveChanges();
        //            Updated = true;
        //        }
        //        else
        //        {
        //            //AdvantageFramework.Navigation.ShowMessageBox(ErrorText);
        //        }
        //    }
        //    catch (Exception ex)
        //    {
        //        Updated = false;
        //    }
        //    finally
        //    {
        //        UpdateRet = Updated;
        //    }

        //    return UpdateRet;
        //}

        //public static bool ClearUserDefinedValue(AdvantageFramework.Core.Database.DbContext DbContext, AdvantageFramework.Core.Database.Entities.UserDefinedLabelTables UserDefinedLabelTable)
        //{
        //    bool ClearUserDefinedValueRet = default;

        //    // objects
        //    bool UserDefinedValueCleared = false;
        //    try
        //    {
        //        if (UserDefinedLabelTable == AdvantageFramework.Core.Database.Entities.UserDefinedLabelTables.JOB_LOG_UDV1)
        //        {
        //            DbContext.Database.ExecuteSqlCommand("UPDATE dbo.JOB_LOG SET UDV1_CODE = NULL");
        //        }
        //        else if (UserDefinedLabelTable == AdvantageFramework.Core.Database.Entities.UserDefinedLabelTables.JOB_LOG_UDV2)
        //        {
        //            DbContext.Database.ExecuteSqlCommand("UPDATE dbo.JOB_LOG SET UDV2_CODE = NULL");
        //        }
        //        else if (UserDefinedLabelTable == AdvantageFramework.Core.Database.Entities.UserDefinedLabelTables.JOB_LOG_UDV3)
        //        {
        //            DbContext.Database.ExecuteSqlCommand("UPDATE dbo.JOB_LOG SET UDV3_CODE = NULL");
        //        }
        //        else if (UserDefinedLabelTable == AdvantageFramework.Core.Database.Entities.UserDefinedLabelTables.JOB_LOG_UDV4)
        //        {
        //            DbContext.Database.ExecuteSqlCommand("UPDATE dbo.JOB_LOG SET UDV4_CODE = NULL");
        //        }
        //        else if (UserDefinedLabelTable == AdvantageFramework.Core.Database.Entities.UserDefinedLabelTables.JOB_LOG_UDV5)
        //        {
        //            DbContext.Database.ExecuteSqlCommand("UPDATE dbo.JOB_LOG SET UDV5_CODE = NULL");
        //        }

        //        UserDefinedValueCleared = true;
        //    }
        //    catch (Exception ex)
        //    {
        //        UserDefinedValueCleared = false;
        //    }
        //    finally
        //    {
        //        ClearUserDefinedValueRet = UserDefinedValueCleared;
        //    }

        //    return ClearUserDefinedValueRet;
        //}

        #endregion
    }
}

File: AdvantageFramework/Database/Procedures/Job.vb
Namespace Database.Procedures.Job

    <HideModuleName()> _
    Public Module Methods

#Region " Constants "



#End Region

#Region " Enum "



#End Region

#Region " Variables "



#End Region

#Region " Properties "



#End Region

#Region " Methods "

#Region "  Core Entities "

        Public Function LoadCore(ByVal Jobs As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Job)) As IEnumerable(Of AdvantageFramework.Database.Core.Job)

            Try

                LoadCore = Jobs _
                           .Select(Function(Entity) New With {Entity.Number,
                                                              Entity.OfficeCode,
                                                              Entity.ClientCode,
                                                              Entity.DivisionCode,
                                                              Entity.ProductCode,
                                                              Entity.Description,
                                                              Entity.IsOpen}) _
                           .Select(Function(Entity) New AdvantageFramework.Database.Core.Job With {.Number = Entity.Number,
                                                                                                   .OfficeCode = Entity.OfficeCode,
                                                                                                   .ClientCode = Entity.ClientCode,
                                                                                                   .DivisionCode = Entity.DivisionCode,
                                                                                                   .ProductCode = Entity.ProductCode,
                                                                                                   .Description = Entity.Description,
                                                                                                   .IsOpen = Entity.IsOpen}).ToList

            Catch ex As Exception
                LoadCore = Nothing
            End Try

        End Function
        Public Function LoadCore(ByVal DbContext As AdvantageFramework.Database.DbContext) As IEnumerable(Of AdvantageFramework.Database.Core.Job)

            Try

                LoadCore = LoadCore(Load(DbContext))

            Catch ex As Exception
                LoadCore = Nothing
            End Try

        End Function

#End Region

        Public Function LoadAllOpenByClientCode(ByVal DbContext As Database.DbContext, ByVal ClientCode As String) As System.Data.Entity.Infrastructure.DbQuery(Of Database.Entities.Job)

            LoadAllOpenByClientCode = From Job In LoadAllOpen(DbContext)
                                      Where Job.ClientCode = ClientCode
                                      Select Job

        End Function
        Public Function LoadByUserForEmployeeCode(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal SecurityDbContext As AdvantageFramework.Security.Database.DbContext, ByVal UserCode As String, ByVal EmployeeCode As String, ByVal OpenByJobProcessNumberOnly As Boolean) As System.Data.Entity.Infrastructure.DbQuery(Of Database.Entities.Job)

            Dim ExcludedJobProcessControlNumbers As Integer() = Nothing
            Dim UserClientDivisionProductAccessList As System.Collections.Generic.List(Of AdvantageFramework.Security.Database.Entities.UserClientDivisionProductAccess) = Nothing
            Dim EmployeeUserCodes As String() = Nothing
            Dim JobNumbers As Integer() = Nothing

            ExcludedJobProcessControlNumbers = New Integer() {2, 5, 7, 10, 11, 6, 12}

            Try

                Try

                    EmployeeUserCodes = (From Entity In AdvantageFramework.Security.Database.Procedures.User.LoadByEmployeeCode(SecurityDbContext, EmployeeCode)
                                         Select [UC] = Entity.UserCode).ToArray

                Catch ex As Exception
                    EmployeeUserCodes = Nothing
                End Try

                If EmployeeUserCodes IsNot Nothing Then

                    If EmployeeUserCodes.Contains(UserCode) Then

                        EmployeeUserCodes = {UserCode}

                    End If

                    Try

                        UserClientDivisionProductAccessList = (From Entity In AdvantageFramework.Security.Database.Procedures.UserClientDivisionProductAccess.Load(SecurityDbContext)
                                                               Where EmployeeUserCodes.Contains(Entity.UserCode)
                                                               Select Entity).ToList

                    Catch ex As Exception
                        UserClientDivisionProductAccessList = Nothing
                    End Try

                End If

                JobNumbers = (From Entity In AdvantageFramework.Database.Procedures.JobComponent.Load(DbContext)
                              Where ExcludedJobProcessControlNumbers.Contains(Entity.JobProcessNumber) = False OrElse
                                        ExcludedJobProcessControlNumbers.Contains(Entity.JobProcessNumber) = If(OpenByJobProcessNumberOnly = False, True, False)
                              Select Entity.JobNumber).Distinct.ToArray

                If UserClientDivisionProductAccessList IsNot Nothing AndAlso UserClientDivisionProductAccessList.Count > 0 Then

                    JobNumbers = (From Entity In (From Job In LoadAllOpen(DbContext)
                                                  Select New With {.ClientCode = Job.ClientCode,
                                                                   .DivisionCode = Job.DivisionCode,
                                                                   .ProductCode = Job.ProductCode,
                                                                   .Number = Job.Number}).ToList
                                  Where UserClientDivisionProductAccessList.Any(Function(UserAccess) UserAccess.ClientCode = Entity.ClientCode AndAlso
                                                                                                    UserAccess.DivisionCode = Entity.DivisionCode AndAlso
                                                                                                    UserAccess.ProductCode = Entity.ProductCode) AndAlso
                                       JobNumbers.Contains(Entity.Number)
                                  Select Entity.Number).ToArray

                End If

                LoadByUserForEmployeeCode = From Job In LoadAllOpen(DbContext)
                                            Where JobNumbers.Contains(Job.Number)
                                            Select Job

            Catch ex As Exception
                LoadByUserForEmployeeCode = Nothing
            End Try

        End Function
        Public Function LoadByUserCode(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal SecurityDbContext As AdvantageFramework.Security.Database.DbContext, ByVal UserCode As String) As System.Data.Entity.Infrastructure.DbQuery(Of Database.Entities.Job)

            Dim UserClientDivisionProductAccess As System.Collections.Generic.List(Of AdvantageFramework.Security.Database.Entities.UserClientDivisionProductAccess) = Nothing
            Dim JobNumbers As Integer() = Nothing
            Dim SQL As String = ""

            Try

                UserClientDivisionProductAccess = AdvantageFramework.Security.Database.Procedures.UserClientDivisionProductAccess.LoadByUserCode(SecurityDbContext, UserCode).ToList

                If UserClientDivisionProductAccess.Count > 0 Then

                    SQL = String.Format("SELECT JOB_NUMBER FROM dbo.JOB_LOG JL " &
                                        "JOIN dbo.SEC_CLIENT SC ON SC.CL_CODE = JL.CL_CODE AND SC.DIV_CODE = JL.DIV_CODE AND SC.PRD_CODE = JL.PRD_CODE " &
                                        "WHERE SC.[USER_ID] = '{0}'", UserCode)

                    Try

                        JobNumbers = DbContext.Database.SqlQuery(Of Integer)(SQL).ToArray

                    Catch ex As Exception
                        JobNumbers = Nothing
                    End Try

                    LoadByUserCode = From Job In DbContext.GetQuery(Of Database.Entities.Job)
                                     Where JobNumbers.Contains(Job.Number)
                                     Select Job

                Else

                    LoadByUserCode = Load(DbContext)

                End If

            Catch ex As Exception
                LoadByUserCode = Nothing
            End Try

        End Function
        Public Function LoadOpenByUserCode(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal SecurityDbContext As AdvantageFramework.Security.Database.DbContext, ByVal UserCode As String) As System.Data.Entity.Infrastructure.DbQuery(Of Database.Entities.Job)

            Dim UserClientDivisionProductAccess As System.Collections.Generic.List(Of AdvantageFramework.Security.Database.Entities.UserClientDivisionProductAccess) = Nothing
            Dim JobNumbers As Integer() = Nothing
            Dim SQL As String = ""

            Try

                UserClientDivisionProductAccess = AdvantageFramework.Security.Database.Procedures.UserClientDivisionProductAccess.LoadByUserCode(SecurityDbContext, UserCode).ToList

                If UserClientDivisionProductAccess.Count > 0 Then

                    SQL = String.Format("SELECT JOB_NUMBER FROM dbo.JOB_LOG JL " &
                                        "JOIN dbo.SEC_CLIENT SC ON SC.CL_CODE = JL.CL_CODE AND SC.DIV_CODE = JL.DIV_CODE AND SC.PRD_CODE = JL.PRD_CODE " &
                                        "WHERE SC.[USER_ID] = '{0}'", UserCode)

                    Try

                        JobNumbers = DbContext.Database.SqlQuery(Of Integer)(SQL).ToArray

                    Catch ex As Exception
                        JobNumbers = Nothing
                    End Try

                    LoadOpenByUserCode = From Job In LoadAllOpen(DbContext)
                                         Where JobNumbers.Contains(Job.Number)
                                         Select Job

                Else

                    LoadOpenByUserCode = LoadAllOpen(DbContext)

                End If

            Catch ex As Exception
                LoadOpenByUserCode = Nothing
            End Try

        End Function
        Public Function LoadByClientCode(ByVal DbContext As Database.DbContext, ByVal ClientCode As String) As System.Data.Entity.Infrastructure.DbQuery(Of Database.Entities.Job)

            LoadByClientCode = From Job In DbContext.GetQuery(Of Database.Entities.Job)
                               Where Job.ClientCode = ClientCode
                               Select Job

        End Function
        Public Function LoadByClientAndDivisionCode(ByVal DbContext As Database.DbContext, ByVal ClientCode As String, ByVal DivisionCode As String) As System.Data.Entity.Infrastructure.DbQuery(Of Database.Entities.Job)

            LoadByClientAndDivisionCode = From Job In DbContext.GetQuery(Of Database.Entities.Job)
                                          Where Job.ClientCode = ClientCode AndAlso
                                                Job.DivisionCode = DivisionCode
                                          Select Job

        End Function
        Public Function LoadByClientAndDivisionAndProductCode(ByVal DbContext As Database.DbContext, ByVal ClientCode As String, ByVal DivisionCode As String, ByVal ProductCode As String) As System.Data.Entity.Infrastructure.DbQuery(Of Database.Entities.Job)

            LoadByClientAndDivisionAndProductCode = From Job In DbContext.GetQuery(Of Database.Entities.Job)
                                                    Where Job.ClientCode = ClientCode AndAlso
                                                          Job.DivisionCode = DivisionCode AndAlso
                                                          Job.ProductCode = ProductCode
                                                    Select Job

        End Function
        Public Function LoadByOfficeCode(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal OfficeCode As String) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Job)

            LoadByOfficeCode = From Job In DbContext.GetQuery(Of Database.Entities.Job)
                               Where Job.OfficeCode = OfficeCode
                               Select Job

        End Function
        Public Function LoadByJobID(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal JobID As Integer) As AdvantageFramework.Database.Entities.Job

            Try

                LoadByJobID = (From Job In DbContext.GetQuery(Of Database.Entities.Job)
                               Where Job.ID = JobID
                               Select Job).SingleOrDefault

            Catch ex As Exception
                LoadByJobID = Nothing
            End Try

        End Function
        Public Function LoadByJobNumber(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal JobNumber As Integer) As AdvantageFramework.Database.Entities.Job

            Try

                LoadByJobNumber = (From Job In DbContext.GetQuery(Of Database.Entities.Job)
                                   Where Job.Number = JobNumber
                                   Select Job).SingleOrDefault

            Catch ex As Exception
                LoadByJobNumber = Nothing
            End Try

        End Function
        Public Function LoadBySalesClassAndFormatCode(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal SalesClassCode As String, ByVal SalesClassFormatCode As String) As System.Data.Entity.Infrastructure.DbQuery(Of Database.Entities.Job)

            Try

                LoadBySalesClassAndFormatCode = (From Job In DbContext.GetQuery(Of Database.Entities.Job)
                                                 Where Job.SalesClassCode = SalesClassCode AndAlso Job.SalesClassFormatCode = SalesClassFormatCode
                                                 Select Job)

            Catch ex As Exception
                LoadBySalesClassAndFormatCode = Nothing
            End Try

        End Function
        Public Function LoadAllOpen(ByVal DbContext As AdvantageFramework.Database.DbContext) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Job)

            LoadAllOpen = From Job In DbContext.GetQuery(Of Database.Entities.Job)
                          Where Job.IsOpen = 1
                          Select Job

        End Function
        Public Function Load(ByVal DbContext As AdvantageFramework.Database.DbContext) As System.Data.Entity.Infrastructure.DbQuery(Of AdvantageFramework.Database.Entities.Job)

            Load = From Job In DbContext.GetQuery(Of Database.Entities.Job)
                   Select Job

        End Function
        Public Function Update(ByVal DbContext As AdvantageFramework.Database.DbContext, ByVal Job As AdvantageFramework.Database.Entities.Job) As Boolean

            'objects
            Dim Updated As Boolean = False
            Dim IsValid As Boolean = True
            Dim ErrorText As String = ""

            Try

                DbContext.UpdateObject(Job)

                ErrorText = Job.ValidateEntity(IsValid)

                If IsValid Then

                    DbContext.SaveChanges()

                    Updated = True

                Else

                    AdvantageFramework.Navigation.ShowMessageBox(ErrorText)

                End If

            Catch ex As Exception
                Updated = False
            Finally
                Update = Updated
            End Try

        End Function
        Public Function ClearUserDefinedValue(ByVal DbContext As AdvantageFramework.Database.DbContext, UserDefinedLabelTable As AdvantageFramework.Database.Entities.UserDefinedLabelTables) As Boolean

            'objects
            Dim UserDefinedValueCleared As Boolean = False

            Try

                If UserDefinedLabelTable = AdvantageFramework.Database.Entities.UserDefinedLabelTables.JOB_LOG_UDV1 Then

                    DbContext.Database.ExecuteSqlCommand("UPDATE dbo.JOB_LOG SET UDV1_CODE = NULL")

                ElseIf UserDefinedLabelTable = AdvantageFramework.Database.Entities.UserDefinedLabelTables.JOB_LOG_UDV2 Then

                    DbContext.Database.ExecuteSqlCommand("UPDATE dbo.JOB_LOG SET UDV2_CODE = NULL")

                ElseIf UserDefinedLabelTable = AdvantageFramework.Database.Entities.UserDefinedLabelTables.JOB_LOG_UDV3 Then

                    DbContext.Database.ExecuteSqlCommand("UPDATE dbo.JOB_LOG SET UDV3_CODE = NULL")

                ElseIf UserDefinedLabelTable = AdvantageFramework.Database.Entities.UserDefinedLabelTables.JOB_LOG_UDV4 Then

                    DbContext.Database.ExecuteSqlCommand("UPDATE dbo.JOB_LOG SET UDV4_CODE = NULL")

                ElseIf UserDefinedLabelTable = AdvantageFramework.Database.Entities.UserDefinedLabelTables.JOB_LOG_UDV5 Then

                    DbContext.Database.ExecuteSqlCommand("UPDATE dbo.JOB_LOG SET UDV5_CODE = NULL")

                End If

                UserDefinedValueCleared = True

            Catch ex As Exception
                UserDefinedValueCleared = False
            Finally
                ClearUserDefinedValue = UserDefinedValueCleared
            End Try

        End Function

#End Region

    End Module

End Namespace

